common/geth/multihoming_client_test.go

```diff
@@ -0,0 +1,157 @@
+	for i := 0; i < numCall; i++ {
+		ctx := context.Background()
+		_, err := client.ChainID(ctx)
+		require.NotNil(t, err)
+	}
+}
+
+func make500Error() rpc.HTTPError {
+	var httpRespError rpc.HTTPError
+	httpRespError.StatusCode = 500
+	httpRespError.Status = "INTERNAL_ERROR"
+	httpRespError.Body = []byte{}
+	return httpRespError
+}
+
+func TestMultihomingClientSenderFaultZeroRetry(t *testing.T) {
+	// 2xx and 4xx attributes to sender's fault, RPC should not rotate
+	statusCodes := []int{201, 202, 400, 401, 403, 429}
+	for _, sc := range statusCodes {
+		var httpRespError rpc.HTTPError
+		httpRespError.StatusCode = sc
+		httpRespError.Status = "INTERNAL_ERROR"
+		httpRespError.Body = []byte{}
+
+		client, _ := makeTestMultihomingClient(0, httpRespError)
+
+		index, _ := client.GetRPCInstance()
+		require.Equal(t, index, 0)
+
+		makeFailureCall(t, client, 1)
+
+		// given num retry is 0, when failure arises above, current rpc should becomes the next one
+		index, _ = client.GetRPCInstance()
+		require.Equal(t, index, 0)
+	}
+}
+
+func TestMultihomingClientRPCFaultZeroRetry(t *testing.T) {
+	httpRespError := make500Error()
+	client, _ := makeTestMultihomingClient(0, httpRespError)
+
+	index, _ := client.GetRPCInstance()
+	require.Equal(t, index, 0)
+
+	makeFailureCall(t, client, 1)
+
+	// given num retry is 0, when failure arises above, current rpc should becomes the next one
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 1)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 2)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 0)
+}
+
+func TestMultihomingClientRPCFaultOneRetry(t *testing.T) {
+	httpRespError := make500Error()
+
+	client, _ := makeTestMultihomingClient(1, httpRespError)
+
+	index, _ := client.GetRPCInstance()
+	require.Equal(t, index, 0)
+
+	makeFailureCall(t, client, 1)
+
+	// given num retry is 1, when failure arises above, two rpc are used, current rpc should becomes 2
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 2)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 1)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 0)
+}
+
+func TestMultihomingClientRPCFaultTwoRetry(t *testing.T) {
+	httpRespError := make500Error()
+	client, _ := makeTestMultihomingClient(2, httpRespError)
+
+	index, _ := client.GetRPCInstance()
+	require.Equal(t, index, 0)
+
+	makeFailureCall(t, client, 1)
+
+	// given num retry is 2, when failure arises above, three rpc are used, current rpc should becomes 0
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 0)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 0)
+
+	makeFailureCall(t, client, 1)
+
+	index, _ = client.GetRPCInstance()
+	require.Equal(t, index, 0)
+}
```

Comment:
We should have test cases where it doesn't change the index if the error isn't RPCFault

Reply:
The "TestMultihomingClientSenderFaultZeroRetry" is the sender's fault, there is no retry

Reply:
"TestMultihomingClientSenderFaultZeroRetry" above tests the case